File: legacy/graphql-java/src/main/java/com/example/demo/service/AuthorService.java
package com.example.demo.service;

import com.example.demo.gql.types.Author;
import com.example.demo.model.AuthorEntity;
import com.example.demo.repository.AuthorRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.Collection;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.function.Function;

@Service
@RequiredArgsConstructor
public class AuthorService {
    private static final Function<AuthorEntity, Author> MAPPER = a -> Author.builder()
            .id(a.id().toString())
            .name(a.name())
            .email(a.email())
            .createdAt(a.createdAt())
            .build();
    final AuthorRepository authors;

    public Author getAuthorById(String id) {
        var authorEntity = this.authors.findById(UUID.fromString(id));
        return Optional.ofNullable(authorEntity)
                .map(MAPPER)
                .orElseThrow(() -> new AuthorNotFoundException(id));
    }

    public List<Author> getAuthorByIdIn(Collection<String> ids) {
        var uuids = ids.stream().map(UUID::fromString).toList();
        var authorEntities = this.authors.findByIdIn(uuids);
        return authorEntities.stream().map(MAPPER).toList();
    }
}

File: dgs-kotlin-co/src/main/kotlin/com/example/demo/service/AuthorService.kt
package com.example.demo.service

import com.example.demo.asGqlType
import com.example.demo.gql.types.Author
import com.example.demo.repository.AuthorRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import org.springframework.stereotype.Service
import java.util.*

@Service
class AuthorService(val authors: AuthorRepository) {

    suspend fun getAuthorById(id: String): Author {
        val author = this.authors.findById(UUID.fromString(id)) ?: throw AuthorNotFoundException(id)
        return author.asGqlType()
    }

    // alternative to use kotlin co `Flow`
    fun getAuthorByIdIn(ids: List<String>): Flow<Author> {
        val uuids = ids.map { UUID.fromString(it) };
        return authors.findAllById(uuids).map { it.asGqlType() }
    }
}